impl/maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java

```diff
@@ -700,11 +709,149 @@ private void initProject(MavenProject project, ModelBuilderResult result) {
+                                            + "Use <source><lang>resources</lang></source> in <sources> for custom resource paths.";
+                            logger.warn(message);
+                            result.getProblemCollector()
+                                    .reportProblem(new org.apache.maven.impl.model.DefaultModelProblem(
+                                            message,
+                                            Severity.WARNING,
+                                            Version.V41,
+                                            project.getModel().getDelegate(),
+                                            -1,
+                                            -1,
+                                            null));
+                        }
+                        logger.debug("Injecting module-aware main resource roots for {} modules", modules.size());
+                        for (String module : modules) {
+                            Path resourcePath = baseDir.resolve("src")
+                                    .resolve(module)
+                                    .resolve(ProjectScope.MAIN.id())
+                                    .resolve("resources");
+                            logger.debug("  - Adding main resource root: {} (module: {})", resourcePath, module);
+                            project.addSourceRoot(
+                                    createModularResourceRoot(baseDir, module, ProjectScope.MAIN, outputDirectory));
+                        }
+                    }
+                } else {
+                    // Classic (non-modular) project
+                    if (hasMainResources) {
+                        // Resources configured via <sources> - already added above
+                        if (!resources.isEmpty()
+                                && !hasOnlySuperPomDefaults(resources, baseDir, ProjectScope.MAIN.id())) {
+                            logger.warn("Legacy <resources> element is ignored because main resources are "
+                                    + "configured via <source><lang>resources</lang></source> in <sources>");
+                        }
+                        logger.debug(
+                                "Main resources configured via <sources> element, ignoring legacy <resources> element");
+                    } else {
+                        // Use legacy <resources> element
+                        logger.debug("Using explicit or default resources ({} resources configured)", resources.size());
+                        for (Resource resource : resources) {
+                            project.addSourceRoot(new DefaultSourceRoot(baseDir, ProjectScope.MAIN, resource));
+                        }
+                    }
                 }
-                for (Resource resource : project.getBuild().getDelegate().getTestResources()) {
-                    project.addSourceRoot(new DefaultSourceRoot(baseDir, ProjectScope.TEST, resource));
+
+                /*
```

Comment:
This block of code for test resources seems very similar to above code for main resources, with only minor differences. Can we share the code in a single method?